File: aGSM4CH/src/coarsen_mesh.m
function [T,P,x,y,c] = coarsen_mesh (T,Ti,P,x,y,c,cri)

% This function coarsen the mesh 
% It searches the triangles subjected to coarsening examination
% the triangles who and whose principal neighbor have child but no grandchild
T(:,1) = 0;
[TC] = search_CT(T,Ti); % get the index of triangles under coarsening examination
while ~isempty(TC)
    while ~isempty(TC)
        
        T1 = TC(1);
        if T(T1,Ti.neighbor(2)) == 0 % then this triangle is boundary
            mean_c = (c(T(T1,Ti.vertex(1)))+c(T(T1,Ti.vertex(2)))+c(T(T1,Ti.vertex(3))))/3.0;
            if (mean_c>cri || mean_c<-cri) && T(T1, Ti.level) > 0  
            % c12 = abs(c(T(T1,Ti.vertex(1)))-c(T(T1,Ti.vertex(2))));
            % c23 = abs(c(T(T1,Ti.vertex(1)))-(c(T(T1,Ti.vertex(2)))+c(T(T1,Ti.vertex(3))))/2)*sqrt(2);
            % c31 = abs(c(T(T1,Ti.vertex(3)))-c(T(T1,Ti.vertex(1))));
            % 
            % if c12<cri && c23<cri && c31<cri 
                               
                % neighbor-related changes
                if T(T1,Ti.neighbor(1))~=0
                    ind_n = T(T(T1,Ti.child(2)),Ti.neighbor(2)); % index of T1's 1st child's principal neighbor
                    T(T1,Ti.neighbor(1)) = ind_n;
                    n_ind = logical(T(ind_n, Ti.neighbor) == T(T1, Ti.child(2))); % find which neighbor of ind_n is T1's 1st child
                    T(ind_n,Ti.neighbor(n_ind)) = T1;
                end
                
                if T(T1,Ti.neighbor(3))~=0
                    ind_n = T(T(T1,Ti.child(3)),Ti.neighbor(2)); % index of T1's 2nd child's principal neighbor
                    T(T1,Ti.neighbor(3)) = ind_n;
                    n_ind = find(T(ind_n,Ti.neighbor)==T(T1,Ti.child(3))); % find which neighbor of ind_n is T1's 2nd child
                    T(ind_n,Ti.neighbor(n_ind)) = T1;
                end
                
                
                P(T(T(T1,Ti.child(2)),Ti.vertex(1))) = 0; % remove the information of the mid-point of long edge
                T(T(T1,Ti.child(2:3)),:) = 0; % remove the information of child information
           
                % child-realted changes
                T(T1,Ti.child) = 0; % has no child now
                if T(T1,Ti.mother) ~= 0 % not pioneer, has mother
                    if T(T(T(T1,Ti.mother),Ti.child(2)),Ti.child(1))==0 && T(T(T(T1,Ti.mother),Ti.child(3)),Ti.child(1))==0
                        T(T(T1,Ti.mother),Ti.grandchild) = 0; % mother has no grandchild now
                    end
                                     
                end
            end
 
            TC(1) = []; % remove the triangle from TC
            T(T1,1) = 1; % indicates that this trinalg has been examined
        else % then, the boundary is not boundary
            T2 = T(T1,Ti.neighbor(2)); index_p = [T1 T2];
            mean_c = (c(T(T1,Ti.vertex(1)))+c(T(T1,Ti.vertex(2)))+c(T(T1,Ti.vertex(3))));
            mean_c = mean_c + (c(T(T2,Ti.vertex(1)))+c(T(T2,Ti.vertex(2)))+c(T(T2,Ti.vertex(3))));
            mean_c = mean_c/6.0;
            if (mean_c>cri || mean_c<-cri) % need coarsened
            
            % c121 = abs(c(T(T1,Ti.vertex(1)))-c(T(T1,Ti.vertex(2))));
            % c231 = abs(c(T(T1,Ti.vertex(1)))-(c(T(T1,Ti.vertex(2)))+c(T(T1,Ti.vertex(3))))/2)*sqrt(2);
            % c311 = abs(c(T(T1,Ti.vertex(3)))-c(T(T1,Ti.vertex(1))));
            % c122 = abs(c(T(T2,Ti.vertex(1)))-c(T(T2,Ti.vertex(2))));
            % c232 = abs(c(T(T2,Ti.vertex(1)))-(c(T(T2,Ti.vertex(2)))+c(T(T2,Ti.vertex(3))))/2)*sqrt(2);
            % c312 = abs(c(T(T2,Ti.vertex(3)))-c(T(T2,Ti.vertex(1))));
            % if c121<cri && c231<cri && c311<cri && c122<cri && c232<cri && c312<cri % need coarsened
                
                % neighbor-related changes
                if T(T1,Ti.neighbor(1))~=0
                    ind_n = T(T(T1,Ti.child(2)),Ti.neighbor(2)); % index of T1's 1st child's principal neighbor
                    T(T1,Ti.neighbor(1)) = ind_n;
                    n_ind = logical(T(ind_n,Ti.neighbor)==T(T1,Ti.child(2))); % find which neighbor of ind_n is T1's 1st child
                    T(ind_n,Ti.neighbor(n_ind)) = T1;
                end
                
                if T(T1,Ti.neighbor(3))~=0
                    ind_n = T(T(T1,Ti.child(3)),Ti.neighbor(2)); % index of T1's 2nd child's principal neighbor
                    T(T1,Ti.neighbor(3)) = ind_n;
                    n_ind = logical(T(ind_n,Ti.neighbor)==T(T1,Ti.child(3))); % find which neighbor of ind_n is T1's 2nd child
                    T(ind_n,Ti.neighbor(n_ind)) = T1;
                end
                
                if T(T2,Ti.neighbor(1))~=0
                    ind_n = T(T(T2,Ti.child(2)),Ti.neighbor(2)); % index of T1's 1st child's principal neighbor
                    T(T2,Ti.neighbor(1)) = ind_n;
                    n_ind = find(T(ind_n,Ti.neighbor)==T(T2,Ti.child(2))); % find which neighbor of ind_n is T1's 1st child
                    T(ind_n,Ti.neighbor(n_ind)) = T2;
                end
                
                if T(T2,Ti.neighbor(3))~=0
                    ind_n = T(T(T2,Ti.child(3)),Ti.neighbor(2)); % index of T2's 2nd child's principal neighbor
                    T(T2,Ti.neighbor(3)) = ind_n;
                    n_ind = find(T(ind_n,Ti.neighbor)==T(T2,Ti.child(3))); % find which neighbor of ind_n is T2's 2nd child
                    T(ind_n,Ti.neighbor(n_ind)) = T2;
                end
                
                P(T(T(T1,Ti.child(2)),Ti.vertex(1))) = 0; % remove the information of the mid-point of long edge
                T(T(T1,Ti.child(2:3)),:) = 0; % remove the information of child information
                T(T(T2,Ti.child(2:3)),:) = 0;% remove the information of child information
                
                % child-realted changes
                T(index_p,Ti.child) = 0; % has no child now
                if T(T1,Ti.mother) ~= 0 % not pioneer, has mother
                    if T(T(T(T1,Ti.mother),Ti.child(2)),Ti.child(1))==0 && T(T(T(T1,Ti.mother),Ti.child(3)),Ti.child(1))==0
                        T(T(T1,Ti.mother),Ti.grandchild) = 0; % mother has no grandchild now
                    end
                    
                    if T(T(T(T2,Ti.mother),Ti.child(2)),Ti.child(1))==0 && T(T(T(T2,Ti.mother),Ti.child(3)),Ti.child(1))==0
                        T(T(T2,Ti.mother),Ti.grandchild) = 0; % mother has no grandchild now
                    end                    
                end
            end
 
            TC([1 find(TC==T2)]) = []; % remove the triangle from TC
            T([T1 T2],1) = 1; % indicates that this trinalg has been examined
        end
    end
    
    [TC] = search_CT(T,Ti);
    
end

%% update the T,P,x,y,c
index_DP = find(P==0); % get the index of P need to be deleted
index_RT = find(T(:,Ti.vertex(1))~=0); % get index of remaining triangles
index_DT = find(T(:,Ti.vertex(1)) == 0); % index of useless triangles
crement_P = zeros(length(P),1);
for i = 1:length(index_DP)
    crement_P(index_DP(i)+1:end) = crement_P(index_DP(i)+1:end) - 1; % crement of P index 
end
x(index_DP) = []; y(index_DP) = []; c(index_DP) = []; % delete the unnecessary mid-points
T(index_RT,Ti.vertex)=T(index_RT,Ti.vertex) + crement_P(T(index_RT,Ti.vertex)); % update the vertex index
P=[1:length(x)];

crement_T = zeros(length(T(:,1)),1);
for i = 1:length(index_DT)
    crement_T(index_DT(i)+1:end) = crement_T(index_DT(i)+1:end) - 1; % crement of triangle index
end
% update the triangle index of mother, sister, child, and neighbor
index_mother=find(T(:,Ti.mother)~=0);
T(index_mother,Ti.mother) = T(index_mother,Ti.mother) + crement_T(T(index_mother,Ti.mother));
index_sister=find(T(:,Ti.sister)~=0);
T(index_sister,Ti.sister) = T(index_sister,Ti.sister) + crement_T(T(index_sister,Ti.sister));
index_child=find(T(:,Ti.child(1))~=0);
T(index_child,Ti.child(2)) = T(index_child,Ti.child(2)) + crement_T(T(index_child,Ti.child(2)));
T(index_child,Ti.child(3)) = T(index_child,Ti.child(3)) + crement_T(T(index_child,Ti.child(3)));
index_neighbor=find(T(:,Ti.neighbor(1))~=0);
T(index_neighbor,Ti.neighbor(1)) = T(index_neighbor,Ti.neighbor(1)) + crement_T(T(index_neighbor,Ti.neighbor(1)));
index_neighbor=find(T(:,Ti.neighbor(2))~=0);
T(index_neighbor,Ti.neighbor(2)) = T(index_neighbor,Ti.neighbor(2)) + crement_T(T(index_neighbor,Ti.neighbor(2)));
index_neighbor=find(T(:,Ti.neighbor(3))~=0);
T(index_neighbor,Ti.neighbor(3)) = T(index_neighbor,Ti.neighbor(3)) + crement_T(T(index_neighbor,Ti.neighbor(3)));
T(index_DT,:) = []; % delete the information of deleted triangles
end

File: AMR4PF/coarsen_mesh.py
def coarsen_mesh(T, Ti, P, x, y, c, cri):
    from search_CT import search_CT
    import numpy as np
    # This function coarsen the mesh
    # It searches the triangles subjected to coarsening examination
    # the triangles who and whose principal neighbor have child but no grandchild
    T[:, 0] = -1
    TC = search_CT(T, Ti)
    counter = 0

    while len(TC) > 0:
        while len(TC) > 0:

            T1 = TC[0]
            if T[T1, Ti.neighbor[1]] == -1: # then this triangle is boundary
                mean_c = (c[T[T1, Ti.vertex[0]]] + c[T[T1, Ti.vertex[1]]] + c[T[T1, Ti.vertex[2]]]) / 3

                if np.abs(mean_c)>cri and T[T1, Ti.level] > 0: # need coarsened
                    # neighbor-related changes

                    if T[T1, Ti.neighbor[0]] != -1:
                        ind_n = T[T[T1, Ti.child[1]], Ti.neighbor[1]] # index of T1's 1st child's principal neighbor
                        T[T1, Ti.neighbor[0]] = ind_n
                        n_ind = np.where((T[ind_n, Ti.neighbor]) == T[T1, Ti.child[1]])[0] # find which neighbor of ind_n is T1's 1st child
                        T[ind_n, Ti.neighbor[n_ind]] = T1

                    if T[T1, Ti.neighbor[2]] != -1:
                        ind_n = T[T[T1, Ti.child[2]], Ti.neighbor[1]]  # index of T1's 2nd child's principal neighbor
                        T[T1, Ti.neighbor[2]] = ind_n
                        n_ind = np.where((T[ind_n, Ti.neighbor]) == T[T1, Ti.child[2]])[0] # find which neighbor of ind_n is T1's 2nd child
                        T[ind_n, Ti.neighbor[n_ind]] = T1

                    P[T[T[T1, Ti.child[1]], Ti.vertex[0]]] = -1 # remove information of the mid-point of long edge
                    T[T[T1, Ti.child[1:3]], :] = -1 # remove the information of child information

                    # child-related changes
                    T[T1, Ti.child] = -1 # has no child now
                    if T[T1, Ti.mother] != -1: # not pioneer, has mother
                        if T[T[T[T1, Ti.mother], Ti.child[1]], Ti.child[0]] == -1 and T[T[T[T1, Ti.mother], Ti.child[2]], Ti.child[0]] == -1:
                            T[T[T1, Ti.mother], Ti.grandchild] = -1 # mother has no grandchild now

                TC = np.delete(TC, 0) # remove the triangle from TC
                T[T1, 0] = 0 # indicate the triangle has been checked
            else: # then, the boundary is not a boundary
                T2 = T[T1, Ti.neighbor[1]]
                mean_c = (c[T[T1, Ti.vertex[0]]] + c[T[T1, Ti.vertex[1]]] + c[T[T1, Ti.vertex[2]]]+c[T[T2, Ti.vertex[0]]] + c[T[T2, Ti.vertex[1]]] + c[T[T2, Ti.vertex[2]]]) / 6

                index_p = [T1, T2]

                if np.abs(mean_c) > cri: # need coarsened

                    # neighbor-related changes
                    if T[T1, Ti.neighbor[0]] != -1:
                        ind_n = T[T[T1, Ti.child[1]], Ti.neighbor[1]]  # index of T1's 1st child's principal neighbor
                        T[T1, Ti.neighbor[0]] = ind_n
                        n_ind = np.where((T[ind_n, Ti.neighbor]) == T[T1, Ti.child[1]])[0]  # find which neighbor of ind_n is T1's 1st child
                        T[ind_n, Ti.neighbor[n_ind]] = T1

                    if T[T1, Ti.neighbor[2]] != -1:
                        ind_n = T[T[T1, Ti.child[2]], Ti.neighbor[1]]  # index of T1's 2nd child's principal neighbor
                        T[T1, Ti.neighbor[2]] = ind_n
                        n_ind = np.where((T[ind_n, Ti.neighbor]) == T[T1, Ti.child[2]])[0]  # find which neighbor of ind_n is T1's 2nd child
                        T[ind_n, Ti.neighbor[n_ind]] = T1

                    if T[T2, Ti.neighbor[0]] != -1:
                        ind_n = T[T[T2, Ti.child[1]], Ti.neighbor[1]] # index of T2's 1st child's principal neighbor
                        T[T2, Ti.neighbor[0]] = ind_n
                        n_ind = np.where((T[ind_n, Ti.neighbor]) == T[T2, Ti.child[1]])[0] # find which neighbor of ind_n is T2's 1st child
                        T[ind_n, Ti.neighbor[n_ind]] = T2

                    if T[T2, Ti.neighbor[2]] != -1:
                        ind_n = T[T[T2, Ti.child[2]], Ti.neighbor[1]]  # index of T2's 2nd child's principal neighbor
                        T[T2, Ti.neighbor[2]] = ind_n
                        n_ind = np.where((T[ind_n, Ti.neighbor]) == T[T2, Ti.child[2]])[0] # find which neighbor of ind_n is T2's 2nd child
                        T[ind_n, Ti.neighbor[n_ind]] = T2

                    P[T[T[T1, Ti.child[1]], Ti.vertex[0]]] = -1  # remove information of the mid-point of long edge
                    T[T[T1, Ti.child[1:3]], :] = -1  # remove the information of child information
                    T[T[T2, Ti.child[1:3]], :] = -1  # remove the information of child information

                    # child-related changes
                    i = 0
                    while i != len(index_p):
                        T[index_p[i], Ti.child] = -1 # has no child now
                        i+=1
                    # T[index_p, Ti.child] = -1  has no child now
                    if T[T1, Ti.mother] != -1:  # not pioneer, has mother
                        if T[T[T[T1, Ti.mother], Ti.child[1]], Ti.child[0]] == -1 and T[T[T[T1, Ti.mother], Ti.child[2]], Ti.child[0]] == -1:
                            T[T[T1, Ti.mother], Ti.grandchild] = -1  # mother has no grandchild now

                        if T[T[T[T2, Ti.mother], Ti.child[1]], Ti.child[0]] == -1 and T[T[T[T2, Ti.mother], Ti.child[2]], Ti.child[0]] == -1:
                            T[T[T2, Ti.mother], Ti.grandchild] = -1  # mother has no grandchild now

                delete_these_indices = np.concatenate(([0], np.where(TC==T2)[0]))
                TC = np.delete(TC, delete_these_indices) # remove the triangle from TC
                T[[T1, T2], 0] = 0 # indicate the triangle has been checked

            counter += 1

            break
        TC = search_CT(T, Ti)

    # update T, P, x, y, c
    index_DP = np.where(P==-1)[0] # get the index of the P that need to be deleted
    index_RT = np.where(T[:, Ti.vertex[0]] != -1)[0] # get index of remaining triangles
    index_DT = np.where(T[:, Ti.vertex[0]] == -1)[0] # get index of useless triangles
    crement_P = np.full(len(P), 0)
    for i in range(len(index_DP)):
        crement_P[(index_DP[i] + 1):] = crement_P[(index_DP[i] + 1):] - 1 # crement of P index

    # delete unecessary mid-points
    x = np.delete(x, index_DP)
    y = np.delete(y, index_DP)
    c = np.delete(c, index_DP)

    T[index_RT[:, None], Ti.vertex] += crement_P[T[index_RT[:, None], Ti.vertex]] # update the vertex index
    P = np.arange(0, len(x))

    crement_T = np.full(len(T),0)
    for i in range(len(index_DT)):
        crement_T[(index_DT[i] + 1):] = crement_T[(index_DT[i] + 1):] - 1 # crement of triangle index

    # update triangle index of mother, sister, child, and neighbor
    index_mother = np.where(T[:, Ti.mother] != -1)[0]
    T[index_mother[:, None], Ti.mother] += crement_T[T[index_mother[:, None], Ti.mother]]
    index_sister = np.where(T[:, Ti.sister] != -1)[0]
    T[index_sister[:, None], Ti.sister] += crement_T[T[index_sister[:, None], Ti.sister]]
    index_child = np.where(T[:, Ti.child[0]] != -1)[0]
    T[index_child[:, None], Ti.child[1]] += crement_T[T[index_child[:, None], Ti.child[1]]]
    T[index_child[:, None], Ti.child[2]] += crement_T[T[index_child[:, None], Ti.child[2]]]
    index_neighbor = np.where(T[:, Ti.neighbor[0]] != -1)[0]
    T[index_neighbor[:, None], Ti.neighbor[0]] += crement_T[T[index_neighbor[:, None], Ti.neighbor[0]]]
    index_neighbor = np.where(T[:, Ti.neighbor[1]] != -1)[0]
    T[index_neighbor[:, None], Ti.neighbor[1]] += crement_T[T[index_neighbor[:, None], Ti.neighbor[1]]]
    index_neighbor = np.where(T[:, Ti.neighbor[2]] != -1)[0]
    T[index_neighbor[:, None], Ti.neighbor[2]] += crement_T[T[index_neighbor[:, None], Ti.neighbor[2]]]

    T = np.delete(T, index_DT, axis=0)

    return T, P, x, y, c